File: App/app/src/main/java/com/cinema/ticket_booking/ui/home/MovieAdapter.java
package com.cinema.ticket_booking.ui.home;

import android.view.*;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.data.model.response.MovieSummary;
import com.cinema.ticket_booking.databinding.ItemMovieGridBinding;
import java.util.List;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.ViewHolder> {

    public interface OnMovieClickListener {
        void onClick(String movieId);
    }

    private final List<MovieSummary> movies;
    private final OnMovieClickListener listener;

    public MovieAdapter(List<MovieSummary> movies, OnMovieClickListener listener) {
        this.movies = movies;
        this.listener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemMovieGridBinding b = ItemMovieGridBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(b);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(movies.get(position));
    }

    @Override
    public int getItemCount() {
        return movies.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemMovieGridBinding b;

        ViewHolder(ItemMovieGridBinding b) {
            super(b.getRoot());
            this.b = b;
        }

        void bind(MovieSummary movie) {
            b.tvTitle.setText(movie.getTitle());
            b.tvRating.setText(String.format("★ %.1f", movie.getAvgRating()));
            b.tvDuration.setText(movie.getDuration() + " phút");
            Glide.with(b.ivPoster.getContext())
                    .load(movie.getPosterUrl())
                    .placeholder(R.drawable.ic_movie_placeholder)
                    .error(R.drawable.ic_movie_placeholder)
                    .centerCrop()
                    .into(b.ivPoster);
            b.getRoot().setOnClickListener(v -> listener.onClick(movie.getId()));
        }
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/home/MovieAdapter.kt
package com.cinema.ticket_booking.ui.home

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.MovieSummary
import com.cinema.ticket_booking.databinding.ItemMovieGridBinding

class MovieAdapter(
    private val movies: List<MovieSummary>,
    private val listener: (String) -> Unit
) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val b = ItemMovieGridBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ViewHolder(b)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(movies[position])
    }

    override fun getItemCount(): Int = movies.size

    inner class ViewHolder(private val b: ItemMovieGridBinding) : RecyclerView.ViewHolder(b.root) {
        fun bind(movie: MovieSummary) {
            b.tvTitle.text = movie.title
            b.tvRating.text = String.format("★ %.1f", movie.avgRating)
            b.tvDuration.text = "${movie.duration} phút"
            Glide.with(b.ivPoster.context)
                .load(movie.posterUrl)
                .placeholder(R.drawable.ic_movie_placeholder)
                .error(R.drawable.ic_movie_placeholder)
                .centerCrop()
                .into(b.ivPoster)
            b.root.setOnClickListener { movie.id?.let { id -> listener(id) } }
        }
    }
}
